### PR TITLE
Test262 integration

### PR DESCRIPTION
**Pull request by Protryon**

### CODE REVIEW

File: src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+			// yield not implemented
+			"src/test/resources/test262/test/language/expressions/assignment/dstr-obj-id-identifier-yield-expr.js",
+			"src/test/resources/test262/test/language/statements/for-in/dstr-obj-id-identifier-yield-expr.js",
+			"src/test/resources/test262/test/language/statements/for-of/dstr-obj-id-identifier-yield-expr.js",
+
+			// shift-java#208
+			"src/test/resources/test262/test/language/global-code/decl-lex-restricted-global.js"
+	));
+
+	private static final HashSet<String> skip = new HashSet<>(Arrays.asList(
```
Comment (bakkot) [inline src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java]: I am morally opposed to skipped tests. They should be xfailed, not skipped. Otherwise the expectations file gets stale as new features are implemented.

File: src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+			return null;
+		}
+		test262CommentBegin += 5;
+		int test262CommentEnd = source.indexOf("---*/", test262CommentBegin);
+		if (test262CommentEnd < 0) {
+			return null;
+		}
+
+		// extract categories -- accounting for various authors indentations.
+		Scanner scanner = new Scanner(source.substring(test262CommentBegin, test262CommentEnd));
```
Comment (bakkot) [inline src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java]: I think we probably just want to depend on a yaml parser.

Comment (Protryon) [reply src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java]: Added snakeyaml

File: src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java (hunk 1 above)
Comment (Protryon) [reply src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java]: Removed all skipped tests in favor of xfailing

File: src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+			"src/test/resources/test262/test/language/expressions/await/await-awaits-thenables.js",
+			"src/test/resources/test262/test/language/expressions/await/await-BindingIdentifier-in-global.js",
+			"src/test/resources/test262/test/language/expressions/await/syntax-await-has-UnaryExpression.js",
+			"src/test/resources/test262/test/language/expressions/await/await-throws-rejections.js",
+			"src/test/resources/test262/test/language/expressions/await/await-in-nested-generator.js",
+			"src/test/resources/test262/test/language/expressions/await/await-awaits-thenables-that-throw.js",
+			"src/test/resources/test262/test/language/expressions/await/syntax-await-has-UnaryExpression-with-MultiplicativeExpression.js",
+			"src/test/resources/test262/test/language/expressions/async-arrow-function/arrow-returns-promise.js",
+
+			// yield not implemented
```
Comment (michaelficarra) [inline src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java]: This isn't true, these are bugs with our handling of `yield` in various contexts.

File: src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+import java.nio.file.Paths;
+import java.util.*;
+
+public class Test262 {
+
+	private static final Yaml yamlParser = new Yaml();
+
+	private static final HashSet<String> xfailParse = new HashSet<>(Arrays.asList(
+			// shift-java#198
+			"src/test/resources/test262/test/annexB/language/statements/for-in/nonstrict-initializer.js",
```
Comment (michaelficarra) [inline src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java]: Can you dedupe the test path? These should all start with `src/test/resources/test262/test` so extract that. This will make it easier in case we want to change the location of our test262 submodule or something.

File: src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+			// exponentiation early errors not implemented
+			"src/test/resources/test262/test/language/expressions/exponentiation/exp-operator-syntax-error-delete-unary-expression-base.js",
+			"src/test/resources/test262/test/language/expressions/exponentiation/exp-operator-syntax-error-typeof-unary-expression-base.js",
+			"src/test/resources/test262/test/language/expressions/exponentiation/exp-operator-syntax-error-logical-not-unary-expression-base.js",
+			"src/test/resources/test262/test/language/expressions/exponentiation/exp-operator-syntax-error-void-unary-expression-base.js",
+			"src/test/resources/test262/test/language/expressions/exponentiation/exp-operator-syntax-error-negate-unary-expression-base.js",
+			"src/test/resources/test262/test/language/expressions/exponentiation/exp-operator-syntax-error-plus-unary-expression-base.js",
+			"src/test/resources/test262/test/language/expressions/exponentiation/exp-operator-syntax-error-bitnot-unary-expression-base.js",
+
+			// yield not implemented
```
Comment (michaelficarra) [inline src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java]: Again, these should just be considered bugs.

File: src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+			"src/test/resources/test262/test/language/expressions/function/params-trailing-comma.js",
+			"src/test/resources/test262/test/language/statements/generators/params-trailing-comma-length.js",
+			"src/test/resources/test262/test/language/statements/generators/params-trailing-comma.js",
+			"src/test/resources/test262/test/language/expressions/generators/params-trailing-comma-length.js",
+			"src/test/resources/test262/test/language/expressions/generators/params-trailing-comma.js",
+
+			// broken test
+			"src/test/resources/test262/test/language/module-code/namespace/internals/set-prototype-of-null.js",
+
+			// shift-java#207
```
Comment (michaelficarra) [inline src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java]: Not in ES2016.

File: src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java [hunk 7]
```diff
@@ -0,0 +1,385 @@
...
+			"src/test/resources/test262/test/language/statements/class/definition/methods-gen-yield-star-after-newline.js",
+			"src/test/resources/test262/test/language/statements/class/definition/methods-gen-yield-weak-binding.js",
+			"src/test/resources/test262/test/language/expressions/object/method-definition/yield-as-logical-or-expression.js",
+			"src/test/resources/test262/test/language/expressions/object/method-definition/yield-weak-binding.js",
+			"src/test/resources/test262/test/language/expressions/object/method-definition/yield-as-parameter.js",
+			"src/test/resources/test262/test/language/expressions/object/method-definition/yield-as-label.js",
+			"src/test/resources/test262/test/language/expressions/object/method-definition/yield-star-after-newline.js",
+			"src/test/resources/test262/test/language/expressions/object/method-definition/yield-as-binding-identifier.js",
+
+			// trailing commas not in ES6
```
Comment (michaelficarra) [inline src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java]: More importantly, not in ES2016, the version supported in this branch.

File: src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java [hunk 8]
```diff
@@ -0,0 +1,385 @@
...
+
+	private static final HashSet<String> xfailParse = new HashSet<>(Arrays.asList(
+			// shift-java#198
+			"src/test/resources/test262/test/annexB/language/statements/for-in/nonstrict-initializer.js",
+			"src/test/resources/test262/test/annexB/language/statements/for-in/bare-initializer.js",
+			"src/test/resources/test262/test/annexB/language/statements/for-in/strict-initializer.js",
+			"src/test/resources/test262/test/annexB/language/statements/for-in/var-objectbindingpattern-initializer.js",
+			"src/test/resources/test262/test/annexB/language/statements/for-in/var-arraybindingpattern-initializer.js",
+
+			// async/await not implemented
```
Comment (michaelficarra) [inline src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java]: Can't we filter these based on feature flag? We should whitelist/blacklist supported/unsupported feature flags.

Comment (Protryon) [reply src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java]: No `async` feature flag exists, the flag is much too broad. Trailing commas, etc, do not have flags or feature flags to accompany them.

Comment (bakkot) [reply src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java]: @Protryon There are `async-functions` and `async-iteration` feature flags which cover most of them (and the exceptions are arguably bugs in the test suite). That's [what shift-parser-js does](https://github.com/shapesecurity/shift-parser-js/blob/f973b067adbb7e72fe3a010596bc50c73a5e6f3c/test/test262/expectations/index.js#L13-L14).

I'm surprised this looks so different from the shift-parser-js tests, in fact. For example, there's [well over a hundred trailing comma tests](https://github.com/shapesecurity/shift-parser-js/blob/f973b067adbb7e72fe3a010596bc50c73a5e6f3c/test/test262/expectations/trailing-function-commas.js#L174), but I only see reference to 14 here.

Comment (Protryon) [reply src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java]: @bakkot `async-functions` covers 4 tests, and `async-iteration` covers none (doesn't exist), at least in the commit currently in the repo. As for all the trailing comma tests, they are running and passing. I suppose that in itself could be a bug in the parser, however IIRC trailing commas are in es2016, just not for function arguments or calls. There is no feature flag for trailing commas, however a hack would be to check if the name contains `trailing-comma`.

Comment (bakkot) [reply src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java]: @Protryon ah, that's because the test262 you're using is two years out of date. It should probably be updated - try ` 72f1cfa`.

(Yes, that will mean pulling in tests for new features, but that's OK. And it matches what shift-parser-js does.)

Comment (bakkot) [reply src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java]: > There is no feature flag for trailing commas, however a hack would be to check if the name contains trailing-comma.

Just list them out manually, [like shift-parser-js](https://github.com/shapesecurity/shift-parser-js/blob/f973b067adbb7e72fe3a010596bc50c73a5e6f3c/test/test262/expectations/trailing-function-commas.js). I would expect this project's test expectations to look almost exactly like shift-parser-js's.

**Comment by bakkot** [comment]
@Protryon Let's roll test262 before landing this.

File: src/main/java/com/shapesecurity/shift/es2016/parser/GenericParser.java [hunk 9]
```diff
@@ -1346,6 +1345,11 @@ protected boolean lookaheadAssignmentExpression() {
 
         if (this.lookahead.type != TokenType.EXP) {
             return left;
+        } else if (left.isLeft()) {
+            Expression expression = left.left().fromJust();
+            if (expression instanceof UnaryExpression) {
+                throw this.createUnexpected(this.lookahead);
```
Comment (bakkot) [inline src/main/java/com/shapesecurity/shift/es2016/parser/GenericParser.java]: This looks to me like it would reject `(-1) ** 2`. Do we have a test covering that case? If not we should add one.

File: src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+				return null;
+			}
+			switch (phase) {
+				case "parse":
+					negativeEnum = Test262Info.Test262Negative.PARSE;
+					break;
+				case "early":
+					negativeEnum = Test262Info.Test262Negative.EARLY;
+					break;
+				default:
```
Comment (bakkot) [inline src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java]: I don't like defaults here. We should throw if we encounter something we don't recognize.

File: src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+			ImmutableList<EarlyError> earlyErrors = EarlyErrorChecker.validate(program);
+			failed = earlyErrors.length > 0;
+			failureReason = new RuntimeException("Early Errors: \n" + earlyErrors.foldLeft((str, error) -> error.message + ", " + str, ""));
+		} catch (JsError e) {
+			failed = true;
+			failureReason = e;
+		}
+		if (xfailed && failed == shouldFail || !xfailed && failed != shouldFail) {
+			if (shouldFail != xfailed) {
+				throw new Test262Exception(info.name, "Parsed and should not have: " + path.toString());
```
Comment (bakkot) [inline src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java]: I would prefer more informative messages. For example, this message does not make it clear if the problem is that the test was xfailed and passed, or was not xfailed but failed.

I think each of the four failure cases (where `[ xfailed, shouldFail, failed ]` takes values `[ false, false, true ]`, `[ false, true, false ]`, `[ true, false, false ]`, `[ true, true, true ]`) warrants its own error message.

File: src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+		}
+	}
+
+	private void runTest(@Nonnull Path root, @Nonnull Path path) throws IOException {
+		if (Files.isDirectory(path) || !path.toString().endsWith(".js")) {
+			return;
+		}
+		String source = new String(Files.readAllBytes(path), StandardCharsets.UTF_8);
+		Test262Info info = extractTest262Info(root.relativize(path).toString(), source);
+		if (info == null) { // parse failure, probably a fixture
```
Comment (bakkot) [inline src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java]: I would prefer we explicitly skip fixtures, and then require the frontmatter to parse. That way we don't accidentally pass a test just because its frontmatter is in a format we don't recognize.

Per [INTERPRETING.md](https://github.com/tc39/test262/blob/dec4e5df7481daf086036a152abb30492dec398a/INTERPRETING.md#modules), fixtures are those which end with `_FIXTURE.js`. We can just ignore those files.

**Review by bakkot — APPROVED** [review]
LGTM modulo comments above.

File: src/main/java/com/shapesecurity/shift/es2016/parser/GenericParser.java (hunk 9 above)
Comment (Protryon) [reply src/main/java/com/shapesecurity/shift/es2016/parser/GenericParser.java]: Test added, passing

**Comment by Protryon** [comment]
@bakkot Addressed comments

File: src/test/java/com/shapesecurity/shift/es2016/parser/expressions/BinaryExpressionTest.java [hunk 13]
```diff
@@ -189,5 +186,7 @@ public void testBinaryExpression() throws JsError {
                 new IdentifierExpression("y")));
 
         testScript("(a, e=0)", new BinaryExpression(new IdentifierExpression("a"), BinaryOperator.Sequence, new AssignmentExpression(new AssignmentTargetIdentifier("e"), new LiteralNumericExpression(0.0))));
+
+        testScript("(-1) ** 2", new BinaryExpression(new UnaryExpression(UnaryOperator.Minus, new LiteralNumericExpression(1.0)), BinaryOperator.Exp, new LiteralNumericExpression(2.0)));
```
Comment (bakkot) [inline src/test/java/com/shapesecurity/shift/es2016/parser/expressions/BinaryExpressionTest.java:191]: Given the approach you've taken, can we also add a test `++a**2` does parse (as `(++a)**2`)?

Comment (Protryon) [reply src/test/java/com/shapesecurity/shift/es2016/parser/expressions/BinaryExpressionTest.java:191]: @bakkot Added
